File: LeetCode/Medium/0279_Perfect_Squares/Solution.java
// 2022-11-22 (tabulation)
class Solution {
    public int numSquares(int n) {
        List<Integer> squares = generateSquares(n);
        int[] tab = new int[n + 1];
        Arrays.fill(tab, -1);
        tab[0] = 0;
        for (int i = 0; i < tab.length; i++) {
            if (tab[i] != -1) {
                
                for (int square : squares) {
                    if (i + square < tab.length) {
                        
                        int next = i + square;
                        if (tab[next] == -1 || tab[next] > tab[i] + 1) {
                            tab[next] = tab[i] + 1;
                        }
                        
                    }
                    
                }
                
                
            }
        }
        return tab[n];
    }
    
    private static List<Integer> generateSquares(int n) {
        List<Integer> squares = new ArrayList<Integer>();
        for (int i = 1; i*i <= n; i++) {
            int square = i*i;
            squares.add(square);
        }
        return squares;
    }
}

File: LeetCode/Medium/0279_Perfect_Squares/Solution.cpp
// 2022-11-22 (dp tabulation)
class Solution {
public:
    int numSquares(int n) {
        /*
        input: n number to break down into perfect squares
        output: number of perfect squares that sum to n

        notes:
        - greedy approach of subtracting greatest square doesn't work (ex. 12)
        - 1 4 9 16 25 36 49 64 81 100 121 144

        approach:
        - generate list of squares
        - initialize int vector of size n+1 (initialized with INT_MAX)
            - represents how many perfect squares to reach that number (represented by the idx)
        - DP tabulation and return n
            - for each index, check if current value + 1 < currentidx+square value, and replace it if so
              for every square in our squares list
        */
        
        vector<int> numSquare(n+1, INT_MAX), squares=generateSquares((int)sqrt(n));
        numSquare[0] = 0;
        for (int i=0; i<=n; i++) {
            for (int square : squares) {
                if (i+square > n) continue;
                numSquare[i+square] = min(numSquare[i]+1, numSquare[i+square]);
            }
        }
        return numSquare[n];
    }

    vector<int> generateSquares(int n) {
        vector<int> output;
        for (int i=n; i>=1; i--) {
            output.push_back(i*i);
        }
        return output;
    }
};